### PR TITLE
scheduler, task_pool, task_proxy refactor

### LINKED ISSUE #1182: "cylc submit" should be able to submit a family
Currently it only does single tasks.  Easy change.

### LINKED ISSUE #1550: Allow state reset of individual message outputs.
As argued in #1551, we should not assume that all message outputs need to be completed when a task finishes.  Therefore we need to able to manually reset the state (completed or not) of  individual message triggers, not just just force them all to completed when the task state is reset to succeeded.

### LINKED ISSUE #2228: suite log server warning
Class `SuiteLogServer` in `lib/cylc/network/https/suite_log_server.py` (only used by gcylc to retrieve suite logs?) contains two calls to methods that don't seem to exist: `self.log.warn()` and `self.log.warning()`.

### PR DESCRIPTION
**Pull request by matthewrmshin**
The main drives of this change are:
* Efficiency.
* Reduce strongly coupled logic where possible.

Some headlines:
* Reduce number of attributes in `cylc.task_proxy` and related objects.
* Reduce line counts in `cylc.scheduler`, `cylc.task_pool` and `cylc.task_proxy`.
* Remove `get_task_proxy`.
  * Replaced by `SuiteConfig.get_taskdef` and constructor of `TaskProxy`.
* Singletons related to job management are now instantiated normally. These include:
  * Batch system manager.
  * Job file writer.
  * Job host initialisation manager.
  * Multi-processing pool for job commands, suite events and task events.
* New module `cylc.task_job_mgr`: job management logic from `cylc.task_pool` and `cylc.task_proxy`.
* New module `cylc.suite_events`: suite events logic from `cylc.scheduler`.
* New module `cylc.task_events_mgr`:
  * Task message and task event handler logic in `cylc.scheulder` and `cylc.task_proxy`.
  * Manage task event handlers: these no longer hang onto their originating task proxy objects. The new logic allows task pool to clean up tasks without having to worry about their incomplete event handlers.
* New module `cylc.suite_db_mgr`:
  * Database management logic from `cylc.scheduler`, `cylc.task_pool` and `cylc.task_proxy`.
  * Manage queueing of database operations - improve efficiency by using a single set of queues.
  * Manage database recovery, clean up, etc.
* Rationalise task state reset logic, by reducing the number of entry points to the `reset_state` and `set_state` methods. Apart from `hold` and `release`, all state change will now call `reset_state`.
* Close #1550. Rewrite `cylc.task_outputs` to allow manual reset of individual outputs to complete/incomplete. (Outputs can be referenced by either the trigger names or by the message strings.)
* Close #1182. `cylc submit` can now submit multiple families and tasks in a single command.

[HJO: Close #2228.]

### CODE REVIEW

**Comment by matthewrmshin** [comment]
Although I have got some more ideas, I think I have done enough here already. I am running the profile battery at the moment. The 1st set of results based on the complex suite is showing promises.

Version | Run           | Elapsed Time (s) | CPU Time - Total (s) | Max Memory (kb)
------- | ------------- | ---------------- | -------------------- | ---------------
HEAD    | complex suite | 1380.7 |           489.2                |         86108.0
7.2.1   | complex suite | 1547.7 |           609.3                |         90024.0

**Comment by matthewrmshin** [comment]
Some other ideas that should be in new issues.

(I'd quite like to have a rethink of our usage of the multi-processing pool. Currently, we start a pool of N child processes before the suite starts, which are kept in memory. This should not be necessary if we can just manage a simple pool of sub-processes - given that all our usages are to run external commands. Not keeping child processes in memory should, in theory, reduces the memory footprint during quiet times while a suite is waiting for events.)

(I'd also like to have a rethink of the poll timer logic. Instead of a poll timer per task, it is probably more efficient to have a poll timer per (host, batch system) combo. If we are polling a task on a given (host, batch system), we might as well poll the other tasks on the same combo as well. On job submission, a task will subscribe its job to the relevant poll timer, which will then manage the poll timing to suit all the subscribed jobs. E.g. if several jobs need to be polled within a small interval of time, they will actually be done together once.)

**Comment by matthewrmshin** [comment]
`cylc run` of the complex suite today (adding 7.0.0 in the mix) gives the following stats:

Version|Run          |Elapsed Time (s)|CPU Time - Total (s)|Max Memory (kb)
-------|-------------|----------------|--------------------|---------------
7.0.0  |complex suite|1510.3          |588.0               |94064.0        
7.2.1  |complex suite|1584.9          |621.8               |89876.0        
HEAD   |complex suite|1378.7          |488.0               |84444.0        

`cylc validate` of the complex suite gives the following stats:

Version|Run          |Elapsed Time (s)|CPU Time - Total (s)|Max Memory (kb)
-------|-------------|----------------|--------------------|---------------
7.0.0  |complex suite|16.5            |16.1                |79780.0        
7.2.1  |complex suite|16.2            |15.9                |79788.0        
HEAD   |complex suite|15.6            |15.4                |78452.0        

All numbers going the right direction for this branch.

**Comment by matthewrmshin** [comment]
(This branch will conflict with #2220, so I'll suggest that we don't merge this until after #2220. I am happy to deal with the conflicts.)

@hjoliver @oliver-sanders please have an initial look of this change.

**Comment by matthewrmshin** [comment]
It may be easier to review the individual changesets. You can expand the `...` to see the purpose of each change.

**Comment by hjoliver** [comment]
> I'll suggest that we don't merge this until after #2220. I am happy to deal with the conflicts.)

Thanks, much appreciated.  I'll finish off 2220 quickly.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by matthewrmshin** [comment]
Branch re-based.

**Comment by matthewrmshin** [comment]
Branch re-based.

File: bin/cylc-validate [hunk 1]
```diff
@@ -90,19 +90,15 @@ def main():
     for name in cfg.taskdefs.keys():
         try:
             itask = TaskProxy(
-                cfg.taskdefs[name],
-                cfg.start_point,
-                is_startup=True,
-                validate_mode=True)
```
Comment (matthewrmshin) [inline bin/cylc-validate]: I think they are unrelated.

File: lib/cylc/scheduler.py [hunk 2]
```diff
@@ -1691,9 +1597,8 @@ def shutdown(self, reason=None):
                 self.task_job_mgr.unlink_hosts_contacts(self.suite)
 
         # disconnect from suite-db, stop db queue
-        if getattr(self, "db", None) is not None:
-            self.pri_dao.close()
-            self.pub_dao.close()
+        if getattr(self, "suite_db_mgr", None) is not None:
```
Comment (oliver-sanders) [inline lib/cylc/scheduler.py]: `if hasattr(self, 'suite_db_mgr):`

Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: No, the attribute is set to None by init. Between initialisation of the scheduler class and initialisation of the suite DB manager class, the attribute may be undefined.

Comment (matthewrmshin) [inline lib/cylc/scheduler.py]: Actually, I lied. The attribute is always set to a value, so I think I can get rid of this test.

**Comment by matthewrmshin** [comment]
Branch re-based + small tweaks.

**Comment by matthewrmshin** [comment]
Branch re-based (to resolve conflicts).

**Comment by matthewrmshin** [comment]
Branch re-based to resolve conflicts.

File: lib/cylc/task_pool.py [hunk 3]
```diff
@@ -822,7 +822,9 @@ def match_dependencies(self):
         """
         all_outputs = {}   # all_outputs[message] = taskid
```
Comment (matthewrmshin) [inline lib/cylc/task_pool.py]: Sorry, this is not possible at the moment, as it has implication all the way down the `satisfy_me` method of the prerequisite objects.

File: bin/cylc-submit [hunk 4]
```diff
@@ -105,26 +115,30 @@ def main():
...
-        get_point(point_str).standardise(),
-        is_startup=True)
+    ret_code = 0
     if options.dry_run:
-        if not task_job_mgr.prep_submit_task_jobs(
-                suite, [itask], dry_run=True):
-            sys.exit(1)
-        print('JOB SCRIPT=%s' % itask.local_job_file_path)
+        task_job_mgr.prep_submit_task_jobs(suite, itasks, dry_run=True)
+        ret_code = 0
```
Comment (matthewrmshin) [inline bin/cylc-submit]: Removed.

**Comment by hjoliver** [comment]
> @hjoliver Could you confirm that your dummy/simulation logic has been preserved.

Confirmed - seems to be OK.

**Comment by hjoliver** [comment]
I'll echo Oliver's comments - the larger-scale logic changes are sensible; I can't see any problems closer-up; performance has improved; and the test battery passes... so I'm inclined to approve and not spend several more days staring at the code ... maybe in future we could aim to split this sort of major refactoring into multiple smaller PRs though? (I'm guilty of equally big ones in the past too, admittedly :-)

**Comment by hjoliver** [comment]
> maybe in future we could aim to split this sort of major refactoring into multiple smaller PRs though

(to be fair, your commits are nicely separated)

File: lib/cylc/task_events_mgr.py [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+from cylc.cfgspec.globalcfg import GLOBAL_CFG
+import cylc.flags
+from cylc.mp_pool import SuiteProcContext
+from cylc.suite_logging import ERR, LOG
+from cylc.task_proxy import TaskProxy
+from cylc.task_state import TASK_STATUS_SUCCEEDED
+
+
+class TaskEventsManager(object):
+    """Task events manager."""
```
Comment (hjoliver) [inline lib/cylc/task_events_mgr.py]: (the doc strings would seem more helpful if the class names were less helpful 😁)

**Comment by oliver-sanders** [comment]
Profiling results of the "diamond" experiment show a good improvement to scaling.

memory
cpu-time

**Pull request merged**